### PR TITLE
Fix-up event_fields in filters

### PR DESCRIPTION
**Pull request by clokep**
The Synapse filtering implementation does a few off-spec things:

* Rejects any backslashes except for `\.`.
* Does not properly unescape multiple backslashes.

This was clarified in MSC3980, but from the comments in Synapse it seems this was always supposed to be supported. You can also see the [next spec release[(https://spec.matrix.org/unstable/appendices/#dot-separated-property-paths).

This essentially ports the implementation from matrix-org/matrix-js-sdk#3134. We could have used a regular expression (probably?), but it just becomes hard to follow and I'm very unconvinced that it is faster.

### CODE REVIEW

File: synapse/events/utils.py [hunk 1]
```diff
@@ -253,14 +245,76 @@ def _copy_field(src: JsonDict, dst: JsonDict, field: List[str]) -> None:
...
+                # Otherwise, just add the current character.
+                part += c
+
+    # Ensure the final part is included. If there's an open escape sequence
+    # it should be included.
+    if escaped:
+        part += "\\"
+    result.append(part)
+
+    return result
```
Comment (squahtx) [inline synapse/events/utils.py:307]: I'm a little sad about the performance here, because the repeated string concatenation takes n\*\*2 time, unless Python is very clever.

If we wanted to, we could track the start and end positions of each part and substring them out. Then, if we saw any escapes, we could do some heavier processing.

However, field names are usually small so the current approach is probably okay.

**Review by squahtx — APPROVED** [review]

File: tests/events/test_utils.py [hunk 2]
```diff
@@ -794,3 +796,40 @@ def test_invalid_types_raise_type_error(self) -> None:
...
+            ["m\\foo", ["m\\foo"]],
+            ["m\\\\foo", ["m\\foo"]],
+            ["m\\\\\\foo", ["m\\\\foo"]],
+            ["m\\\\\\\\foo", ["m\\\\foo"]],
+            # Ensure that escapes at the end don't cause issues.
+            ["m.foo\\", ["m", "foo\\"]],
+            ["m.foo\\\\", ["m", "foo\\"]],
+            ["m.foo\\.", ["m", "foo."]],
+            ["m.foo\\\\.", ["m", "foo\\", ""]],
+            ["m.foo\\\\\\.", ["m", "foo\\."]],
```
[the remark this answers is not in the document]
Comment (clokep) [reply tests/events/test_utils.py:823]: Oh good call, I copied this from the matrix-js-sdk which doesn't have that. 👍

File: synapse/events/utils.py (hunk 1 above)
Comment (squahtx) [reply synapse/events/utils.py:307]: Building up a list containing chars to be concatenated all at once would be more optimal (in theory) than concatenating one char at a time. It might be best to benchmark the approaches to make sure I'm not talking nonsense though.

Comment (DMRobertson) [reply synapse/events/utils.py:307]: (On string concatenation: you could try https://docs.python.org/3/library/io.html?highlight=stringio#io.StringIO which I think is more clever than just `+=`ing to a string.)

Comment (clokep) [reply synapse/events/utils.py]: I rewrote it to be quite different, let me know what you think.

File: tests/events/test_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (squahtx) [reply tests/events/test_utils.py]: That's really surprising, oh well.

**Review by squahtx — APPROVED** [review]
The new approach to splitting looks sound to me!

File: synapse/events/utils.py (hunk 1 above)
Comment (squahtx) [reply synapse/events/utils.py]: That's a nice way of doing the splitting. Looks right as far as I can tell.

Comment (clokep) [reply synapse/events/utils.py]: It took me a while to figure out a regular expression method that actually did what I needed...

I suppose I should compile the regexp before merging.
